### PR TITLE
feat(FieldRequirements): add new component

### PR DESCRIPTION
**Pull request by ozgurkececioglu**
[FX-2525]

### Description

- FieldRequirements: added a new component
- forms/PasswordInput: now supporting FieldRequirements as default but it can be hidden with `hideRequirements` (ex: confirm password field)
- I had to touch `FormField` and `FieldWrapper` so please share your considerations for the future

### How to test

- PasswordInput: https://picasso.toptal.net/fx-2525-add-strength-validations-section-to-passwordinput/?path=/story/forms-passwordinput--passwordinput
- FieldRequirements: https://picasso.toptal.net/fx-2525-add-strength-validations-section-to-passwordinput/?path=/story/components-fieldrequirements--fieldrequirements
- Form version(bottom of the page): https://picasso.toptal.net/fx-2525-add-strength-validations-section-to-passwordinput/?path=/story/picasso-forms-form--form

### Screenshots

Screen Shot 2022-03-01 at 10 26 32
Screen Shot 2022-03-01 at 10 26 46
Screen Shot 2022-03-01 at 10 27 02

### Review

- [x] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [x] Annotate all `props` in component with documentation
- [x] Create `examples` for component
- [x] Ensure that deployed demo has expected results and good examples
- [x] Ensure that tests pass by running `yarn test`
- [x] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

[FX-2525]: https://toptal-core.atlassian.net/browse/FX-2525?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/picasso/src/FormField/FormField.tsx [hunk 1]
```diff
@@ -37,6 +47,7 @@ export const FormField = forwardRef<HTMLDivElement, Props>(function FormField(
       {children}
       {error && <FormError className={classes.error}>{error}</FormError>}
       {hint && <FormHint className={classes.hint}>{hint}</FormHint>}
+      {fieldRequirements}
```
Comment (ozgurkececioglu) [inline packages/picasso/src/FormField/FormField.tsx:50]: To achieve placing the field-requirements below our generic field error, I had to do this.

File: packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx [hunk 2]
```diff
@@ -51,6 +51,10 @@ export type Props<
     hideLabelRequiredDecoration?: boolean
     fieldType?: string
     children: (props: any) => React.ReactNode
+    renderFieldRequirements?: (props: {
```
Comment (ozgurkececioglu) [inline packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx:54]: This has to be like that to get the latest value and error state from form-context.

File: packages/picasso-forms/src/PasswordInput/PasswordInput.tsx [hunk 3]
```diff
@@ -1,18 +1,110 @@
...
+          message: '1 special character',
+          validator: passwordValidators.atLeastOneSpecialCharacter
+        },
+        {
+          message: '1 lowercase character',
+          validator: passwordValidators.atLeastOneLowerCaseCharacter
+        }
+      ]}
+    />
+  )
```
Comment (dmaklygin) [inline packages/picasso-forms/src/PasswordInput/PasswordInput.tsx:78]: :+1:

File: packages/picasso/src/FieldRequirements/types.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+export interface FieldRequirement {
+  message: string
+  validator: (value: any) => boolean
```
Comment (dmaklygin) [inline packages/picasso/src/FieldRequirements/types.ts:3]: I'm thinking about possible types of `value`. Maybe we could be more specific about possible types?

File: packages/picasso/src/FieldRequirements/story/Default.example.tsx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+import type { FieldRequirement } from '@toptal/picasso'
+
+const requirements: FieldRequirement[] = [
+  {
+    message: 'Min 5 characters',
+    validator: value => value.length >= 5
+  },
+  {
+    message: 'Max 10 characters',
+    validator: value => value.length < 10
```
Comment (dmaklygin) [inline packages/picasso/src/FieldRequirements/story/Default.example.tsx:12]: shouldn't it be: `value.length <= 10` ?

**Review by dmaklygin — APPROVED** [review]
Great job!

File: packages/picasso/src/FieldRequirements/story/Default.example.tsx (hunk 5 above)
Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/story/Default.example.tsx:12]: Copilot mistake 🙂 I'll fix it.

File: packages/picasso/src/FieldRequirements/types.ts (hunk 4 above)
Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/types.ts:3]: There is a perfect match [ValueType](https://github.com//toptal/picasso/blob/f3da7ae3a8bc3aad1ea67ff8ad81ce2de45aa989/packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx#L23) but it is not exported. And most other places also use `any`. I would prefer exporting ValueType and using it around the repo. WDYT?

**Comment by vvmarulin** [comment]
@ozgurkececioglu 

Checked the demo. Everything works great except for one small thing. When I tried to click submit button after I entered the password (focus on password field), the button wasn't clicked because there is an animation. That is a UX problem because it's unclear from the user's perspective that the button hasn't been clicked.

https://user-images.githubusercontent.com/3861498/156132937-b1a2cf48-8fb1-4972-b148-51e83ff4df66.mp4

You can see on the video, that after the first click on the submit button nothing happened. And only after the second click a validation error for the second field was shown.

Maybe we need some delay for animation so the button doesn't move immediately? WDYT?

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          color='inherit'
+          data-testid={`${dataTestId}-valid-icon`}
+        />
+      ) : error ? (
+        <CloseMinor16 color='red' data-testid={`${dataTestId}-error-icon`} />
+      ) : (
+        <Bullet16
+          color='dark-grey'
+          data-testid={`${dataTestId}-default-icon`}
+        />
```
Comment (OndrejTuma) [inline packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:46]: I'm wondering in which case this can happen. Because to me it seems that either valid or error state are possible. I can only think of some validation that is somewhat informative only (does not reach a valid state) but is not considered in the `validate` function of FieldWrapper. And that seems like a hell of an edgecase 🙂

Comment (vvmarulin) [reply packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:46]: When you focus on the password field for the first time it will show a list with bullet icons. Also while you typing a new value for the first time the requirements will either have a bullet icon or check icon.

Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:46]: @OndrejTuma according to our designs here: https://www.figma.com/file/q2nvjiyO2CLqBv4DeJnU3i/Product-Library-Documentation?node-id=702%3A11760 there is a default state that we are showing bullets only as @vvmarulin explained perfectly.

File: packages/picasso-forms/src/PasswordInput/PasswordInput.tsx [hunk 7]
```diff
@@ -1,18 +1,115 @@
...
+          inputProps.onFocus?.(event)
+
+          setShowContent(true)
+        }
+
+        const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
+          inputProps.onBlur?.(event)
+
+          // Hide the requirements after a short delay
+          setTimeout(() => {
```
Comment (ozgurkececioglu) [inline packages/picasso-forms/src/PasswordInput/PasswordInput.tsx:101]: According to our decision, there should be a delay to start hiding animation for the requirements after onBlur. `transitionDelay` didn't work so I had to provide this solution.

Another possibility is that I can implement my own transition animation with pure CSS, rather than using MUI's Collapse component.

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx (hunk 6 above)
Comment (OndrejTuma) [reply packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:46]: Oh, I missed that 🙇  thanks for explanation

**Review by OndrejTuma — APPROVED** [review]
GJ 👍

**Review by vvmarulin — APPROVED** [review]

File: .changeset/dirty-points-applaud.md [hunk 8]
```diff
@@ -0,0 +1,13 @@
+---
+'@toptal/picasso-forms': minor
+---
+
+---
+### FieldWrapper
+
+- add new prop to render field requirements if exist
```
Comment (separatio) [inline .changeset/dirty-points-applaud.md:8]: I'd rephrase this to: `- add new prop to render passed field requirements`

File: packages/picasso/src/FieldRequirements/FieldRequirements.tsx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+          ))}
+        </Grid>
+      </Collapse>
+    )
+  }
+)
+
+FieldRequirements.defaultProps = {
+  open: false,
+  timeout: 500
```
Comment (TomasSlama) [inline packages/picasso/src/FieldRequirements/FieldRequirements.tsx:93]: is this related to the `TIMEOUT` const in the PasswordInput file? maybe we can reuse it?

File: packages/picasso/src/FieldRequirements/styles.ts [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+export default ({ palette, spacing, sizes: { input } }: Theme) =>
+  createStyles({
+    root: {
+      display: 'flex',
+      width: input.width
+    },
+    description: {
+      marginTop: '0.4rem'
+    },
+    fieldRequirementItem: {
```
Comment (TomasSlama) [inline packages/picasso/src/FieldRequirements/styles.ts:12]: maybe I'm blind, but I didn't find usage of this class?

File: packages/picasso-forms/src/PasswordInput/validators.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  return /[!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]/.test(value)
+}
+
+export default {
+  atLeastEightCharacters,
+  atLeastOneUpperCaseCharacter,
+  atLeastOneLowerCaseCharacter,
+  atLeastOneNumber,
+  atLeastOneSpecialCharacter
+}
```
Comment (TomasSlama) [inline packages/picasso-forms/src/PasswordInput/validators.ts:28]: Should we unit test those validators?

File: packages/picasso-forms/src/PasswordInput/PasswordInput.tsx [hunk 12]
```diff
@@ -1,18 +1,115 @@
...
+  const [showContent, setShowContent] = useState(false)
+
+  const validatePassword: FieldValidator<PasswordInputProps['value']> = (
+    value?: string
+  ) => {
+    if (!value || hideRequirements) {
+      return undefined
+    }
+
+    const validationResult =
```
[the remark this answers is not in the document]
Comment (ozgurkececioglu) [reply packages/picasso-forms/src/PasswordInput/PasswordInput.tsx:34]: I was planning something different but forgot to update this after changing my mind. Good catch!

**Review by augustobmoura — CHANGES_REQUESTED** [review]

File: packages/picasso/src/FieldRequirements/styles.ts (hunk 10 above)
Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/styles.ts:12]: It is being used inside `FieldRequirementItem`. see [here](https://github.com//toptal/picasso/blob/c157f45afe2907210b0ba8840c52f4e950c468c8/packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx#L39)

File: packages/picasso/src/FieldRequirements/types.ts (hunk 4 above)
Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/types.ts]: I created an identical clone type into FieldRequirements and started using it as generic base type 👍

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+      ) : status === 'error' ? (
+        <CloseMinor16
+          color={colorMap[status]}
+          data-testid={`${dataTestId}-error-icon`}
+        />
+      ) : (
+        <Bullet16
+          color={colorMap[status]}
+          data-testid={`${dataTestId}-default-icon`}
+        />
```
Comment (TomasSlama) [inline packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:56]:
```suggestion
      const Icons = {
        error: CloseMinor16,
        // ..
      }
      const IconComponent = Icons[status]
```

and also I think we should not hard code test id `data-testid={`${dataTestId}-valid-icon`}`. So I would suggest `testIds.validIcon`

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
+import React, { PropsWithChildren } from 'react'
+import { makeStyles, Theme } from '@material-ui/core/styles'
+
+import Bullet16 from '../Icon/Bullet16'
+import Typography from '../Typography'
+import Grid from '../Grid'
+import styles from './styles'
+import { CheckMinor16, CloseMinor16 } from '../Icon'
```
Comment (TomasSlama) [inline packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:8]:
```suggestion
import Typography from '../Typography'
import Grid from '../Grid'
import styles from './styles'
import { CheckMinor16, CloseMinor16, Bullet16 } from '../Icon'
```

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+
+  return (
+    <Grid.Item
+      small={6}
+      className={classes.fieldRequirementItem}
+      data-testid={dataTestId}
+    >
+      <IconComponent
+        color={colorMap[status]}
+        data-testid={`${dataTestId}-${status}-icon`}
```
Comment (TomasSlama) [inline packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:52]: we should not have any hardcoded data test id in Picasso. `testIds: { successIcon: string, ...}`

File: packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx (hunk 14 above)
Comment (ozgurkececioglu) [reply packages/picasso/src/FieldRequirements/FieldRequirementItem.tsx:8]: I'm so use to clean these imports with some shortcuts but here, we are not using I guess 🙂 thanks 👍

File: packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx [hunk 16]
```diff
@@ -51,6 +51,10 @@ export type Props<
...
+      value?: ValueType
```
Comment (augustobmoura) [inline packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx:55]: Shouldn't this type be `TInputValue`? This value would be the same as the input right? So in a number input a number, in a password a string, etc

File: packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx [hunk 17]
```diff
@@ -51,6 +51,10 @@ export type Props<
...
+      error?: boolean
+    }) => React.ReactNode
```
[the remark this answers is not in the document]
Comment (augustobmoura) [reply packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx:57]: I don't know if I understood it correctly. I'm not seeing the difference between a function returning a ReactNode and a common component (in fact, FunctionComponents have essentially the same signature as this prop).

Why would calling a function that returns a new ReactNode not trigger a react re-render, but directly creating a ReactNode would do it?

File: packages/picasso-forms/src/PasswordInput/PasswordInput.tsx [hunk 18]
```diff
@@ -1,18 +1,120 @@
...
+
+  const renderFieldRequirements: ({
+    value,
+    error
+  }: {
+    value?: FieldRequirementValueType
+    error?: boolean
+  }) => ReactNode = ({ value, error }) => (
+    <Form.FieldRequirements<string>
+      value={(value ?? '') as string}
```
Comment (augustobmoura) [inline packages/picasso-forms/src/PasswordInput/PasswordInput.tsx:57]: Why not type `value` as a `string` instead of type asserting it? Passwords will always be a `string` (at least for now)

If the problem is that `FieldWrapper` on accepts an non-generic `value`, refer to my comment on `FieldWrapper.tsx` about using `TInputValue` as the type of value

File: packages/picasso/src/FieldRequirements/FieldRequirements.tsx [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+import styles from './styles'
+import FieldRequirementItem, {
+  FieldRequirementItemStatus
+} from './FieldRequirementItem'
+import Typography from '../Typography'
+import { FieldRequirement, ValueType } from './types'
+import Grid from '../Grid'
+
+export interface Props<TInputValue extends ValueType = ValueType>
```
Comment (augustobmoura) [inline packages/picasso/src/FieldRequirements/FieldRequirements.tsx:14]: I don't think that this should have a default type, it's better to require the developer to specify it. Being too generic of type can lead to type assertion hacks or untyped code.

FieldWrapper for example, requires it to be defined by the developer https://github.com/toptal/picasso/blob/510c31766197cce7df7191cb8c27df4b967920e0/packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx#L36

File: packages/picasso/src/FieldRequirements/FieldRequirements.tsx [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    gridContainer?: string
+  }
+}
+
+const ANIMATION_TIMEOUT = 500
+
+const useStyles = makeStyles<Theme>(styles)
+
+export const FieldRequirements = function <
+  TInputValue extends ValueType = ValueType
```
Comment (augustobmoura) [inline packages/picasso/src/FieldRequirements/FieldRequirements.tsx:40]: I don't think that this should have a default type, same argument as in the first comment on `FieldRequirements.tsx`

File: packages/picasso/src/FieldRequirements/index.ts [hunk 21]
```diff
@@ -0,0 +1,10 @@
+import { OmitInternalProps } from '@toptal/picasso-shared'
+
+import { Props } from './FieldRequirements'
+import { ValueType } from './types'
+
+export { default } from './FieldRequirements'
+export * from './types'
+export type FieldRequirementsProps<
+  TInputType extends ValueType = ValueType
```
Comment (augustobmoura) [inline packages/picasso/src/FieldRequirements/index.ts:9]: I don't think that this should have a default type, same argument as in the first comment on `FieldRequirements.tsx`

File: packages/picasso/src/FieldRequirements/types.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  | number
+  | boolean
+  | null
+  | undefined
+  | FileUpload[]
+  | DateOrDateRangeType
+  | Item
+  | Item[]
+
+export interface FieldRequirement<TInputType extends ValueType = ValueType> {
```
Comment (augustobmoura) [inline packages/picasso/src/FieldRequirements/types.ts:18]: I don't think that this should have a default type, same argument as in the first comment on `FieldRequirements.tsx`

**Review by augustobmoura — CHANGES_REQUESTED** [review]

File: packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (ozgurkececioglu) [reply packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx]: Your logic is totally fine. I like it. I've just not been able to find a solution to pass the same component result to `FieldWrapper`, before passing it as `ReactNode` to `FormField`. I've created a variant branch to talk with more solid examples here: https://github.com/toptal/picasso/tree/fx-2525-add-strength-validations-section-to-passwordinput-variant

To achieve your suggestion, as you can see [here](https://github.com/toptal/picasso/blob/4f3d761e4296be3d817121333d6df2624a9d607c/packages/picasso-forms/src/PasswordInput/PasswordInput.tsx#L54), I had to add one more wrapper to pass `showContent` to my `FieldRequirementsComponent` to start and stop the animation. If you run storybook for this variant branch, you would see that `FieldRequirementsComponent` is being recreated after each focusing and bluring. Even if we accept that, because of the recreation, animations couldn't work.

With the render function solution, I've just postponed the return phase to `FieldWrapper's render` and it solved the issue.

If you assist on this variant branch with a working solution, I would definitely be happy and accept that 🙂

**Comment by separatio** [comment]
Didn't finish manually testing, but please pay attention to min-maxing. I have exactly 10 characters written, and the field is red.
image

**Comment by separatio** [comment]
Haven't checked designs yet, but it feels weird that the field is not red while it's clearly failing validation.
image

**Comment by ozgurkececioglu** [comment]
> Haven't checked designs yet, but it feels weird that the field is not red while it's clearly failing validation. image

Again, it is about our input fields globally. I've just used a very simple one to accommodate the validations running. You can check others from: https://picasso.toptal.net/?path=/story/forms-input--input

I think in focus, none of our fields provide an error signal.

File: packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (augustobmoura) [reply packages/picasso-forms/src/FieldWrapper/FieldWrapper.tsx]: Hmmm, I think I understand the problem now. `FieldRequirements` needs an intermediate value, so we create an anonymous component to wrap it up.

The difference between a simple function and a component, in this case, is that when using the JSX syntax the function identity is used on reconciliation. Because we are creating an anonymous component every time, the reconciliation fails every time.

Maybe there's a way to work around that, but I don't think the effort is worth it. We can keep it as a function then. Thanks for writing the example repo for illustration!

**Review by augustobmoura — APPROVED** [review]
